vike/node/plugin/shared/readGitignore/index.ts

```diff
@@ -0,0 +1,90 @@
+export { getGitignoreLines, readAndSetGitignoreLines, parseGitignore }
+
+import { readFile, readdir } from 'node:fs/promises'
+import { join } from 'node:path'
+import { getGlobalObject, unique } from '../../utils.js'
+
+const globalObject = getGlobalObject<{
+  lines: string[]
+  globs: { include: string[]; exclude: string[] }
+}>('readGitignore.ts', { lines: [], globs: { include: [], exclude: [] } })
+
+function getGitignoreLines() {
+  return globalObject
+}
+
+async function readAndSetGitignoreLines(userRootDir: string) {
+  const res = await readdir(userRootDir)
+  const gitignoreFile = res.find((fname) => fname.toLowerCase() === '.gitignore')
+  if (gitignoreFile) {
+    const content = await readFile(join(userRootDir, gitignoreFile), 'utf8')
+    const result = parseGitignore(content)
+    globalObject.lines = result.lines
+    globalObject.globs = result.globs
+    return result
+  }
+}
+
+function parseGitignore(content: string) {
+  const lines = content
+    .replace(/\r\n/g, '\n')
+    .split('\n')
+    .map((line) => line.trim())
+    .filter((line) => line.length && !line.startsWith('#'))
+
+  const uniqueLines = unique(lines)
+  const includeLines: string[] = []
+  const excludeLines: string[] = []
+  function addLine(line: string, include: boolean) {
+    if (include) {
+      includeLines.push(line)
+    } else {
+      excludeLines.push(line)
+    }
+  }
+  for (let line of uniqueLines) {
+    let include = false
+    if (line.startsWith('!')) {
+      line = line.substring(1)
+      include = true
+    }
+
+    const slashPosBefore = line.indexOf('/')
+
+    if (line.startsWith('/')) {
+      line = line.substring(1)
+    }
+
+    if (line.endsWith('/')) {
+      line = `${line}**`
+    }
+
+    const slashPosAfter = line.indexOf('/')
+    if (slashPosAfter === -1) {
+      const startsWithAsterisk = line.startsWith('*')
+      const endWithAsterisk = line.endsWith('*')
+      while (line.startsWith('*')) line = line.substring(1)
```

Comment:
We remove all leading `*` here, right? That means we transform `*foo` into `foo`, and `startsWithAsterisk` is `true`, so later on line 78 we will turn it into `**/*foo*`, which is not the same, right? a `barfoobaz` file wouldn't have matched the original `*foo` expression, but now it will match the new `**/*foo*` expression.

I suspect we can come up with other problems here and I think Rom's idea of using an already existing tool for this instead feels safer. He suggested for example `git ls-files` but I discovered also `git status --ignored`. I have played a bit with it and it seems to be exactly what we want. See also https://stackoverflow.com/questions/466764/git-command-to-show-which-specific-files-are-ignored-by-gitignore

It seems to cover all the edge cases we have mentioned so far, like having .gitignore in ancestor folders or child folders. It won't yield any star or any exclamations marks. It may yield lines starting with `../` if the `.gitignore` comes from an ancestor folder, and all we have to do is skip them. We could then be quite confident to have a working implementation. What do you think?

Reply:
`git status --ignored [--porcelain]` isn't perfect either I just discovered, sorry I stand corrected, but `git ls-files` instead of `glob()` as suggested by Rom seems to be a viable option indeed